convex/agent/memory.ts

```diff
@@ -330,6 +331,155 @@ export const insertMemory = internalMutation({
   },
 });
 
+export const insertReflectionMemories = internalMutation({
+  args: {
+    agentId: v.id('agents'),
+    generationNumber: v.number(),
+
+    reflections: v.array(
+      v.object({
+        description: v.string(),
+        relatedMemoryIds: v.array(v.id('memories')),
+        importance: v.number(),
+        embedding: v.array(v.float64()),
+      }),
+    ),
+  },
+  handler: async (ctx, { agentId, generationNumber, reflections }) => {
+    const agent = await ctx.db.get(agentId);
+    if (!agent) {
+      throw new Error(`Agent ${agentId} not found`);
+    }
+    if (agent.generationNumber !== generationNumber) {
+      throw new Error(
+        `Agent ${agentId} generation number ${agent.generationNumber} does not match ${generationNumber}`,
+      );
+    }
+    // Clear the `isThinking` flag atomically with inserting the memory.
+    await ctx.db.patch(agentId, { isThinking: undefined });
+    const lastAccess = Date.now();
+    for (const { embedding, relatedMemoryIds, ...rest } of reflections) {
+      const embeddingId = await ctx.db.insert('memoryEmbeddings', {
+        playerId: agent.playerId,
+        embedding: embedding,
+      });
+      await ctx.db.insert('memories', {
+        playerId: agent.playerId,
+        embeddingId,
+        lastAccess,
+        ...rest,
+        data: {
+          type: 'reflection',
+          relatedMemoryIds,
+        },
+      });
+    }
+  },
+});
+
+export async function reflectOnMemories(
+  ctx: ActionCtx,
+  agentId: Id<'agents'>,
+  generationNumber: number,
+  playerId: Id<'players'>,
+) {
+  const { memories, lastReflectionTs, name } = await ctx.runQuery(
+    internal.agent.memory.getReflectionMemories,
+    {
+      playerId,
+      numberOfItems: 100,
+    },
+  );
+
+  // should only reflect if lastest 100 items have importance score of >500
+  const sumOfImportanceScore = memories
+    .filter((m) => m._creationTime > (lastReflectionTs ?? 0))
+    .reduce((acc, curr) => acc + curr.importance, 0);
+  const shouldReflect = sumOfImportanceScore > 500;
```

Comment:
I wonder if this is going to get expensive for database bandwidth => we may need to keep this sum denormalized in agent state... until convex supports aggregates of course :)

Reply:
My hope is that we can keep it naive for a while - it should only happen once per conversation, not once per message or step or anything. However, we could keep it in state if we end up with a lot of memories being formed otherwise - e.g. doing activities